oss-internship-2020/libtiff/example/main_sandboxed.cc

```diff
@@ -0,0 +1,305 @@
+  }
+
+  return sandbox2::file::JoinPath(project_path, "test", "images", filename);
+}
+
+absl::Status LibTIFFMain(const std::string& srcfile) {
+  // without addDir to sandbox. to add dir use
+  // sandbox(absolute_path_to_dir, srcfile) or
+  // sandbox(absolute_path_to_dir). file and dir should be exists.
+  // srcfile must also be absolute_path_to_file
+  TiffSapiSandbox sandbox("", srcfile);
+
+  // initialize sapi vars after constructing TiffSapiSandbox
+  sapi::v::UShort h, v;
+  sapi::StatusOr<TIFF*> status_or_tif;
+  sapi::StatusOr<int> status_or_int;
+  sapi::StatusOr<tmsize_t> status_or_long;
+  absl::Status status;
+
+  status = sandbox.Init();
+
+  SAPI_RETURN_IF_ERROR(sandbox.Init());
+
+  TiffApi api(&sandbox);
+  sapi::v::ConstCStr srcfile_var(srcfile.c_str());
+  sapi::v::ConstCStr r_var("r");
+
+  SAPI_ASSIGN_OR_RETURN(
+      status_or_tif, api.TIFFOpen(srcfile_var.PtrBefore(), r_var.PtrBefore()));
+
+  sapi::v::RemotePtr tif(status_or_tif.value());
+  if (!tif.GetValue()) {
+    // tif is NULL
+    return absl::InternalError(absl::StrCat("Could not open ", srcfile));
+  }
+
+  SAPI_ASSIGN_OR_RETURN(status_or_int,
+                        api.TIFFGetField2(&tif, TIFFTAG_YCBCRSUBSAMPLING,
+                                          h.PtrBoth(), v.PtrBoth()));
+  if (status_or_int.value() == 0 || h.GetValue() != 2 || v.GetValue() != 2) {
+    return absl::InternalError("Could not retrieve subsampling tag");
+  }
+
+  SAPI_ASSIGN_OR_RETURN(status_or_long, api.TIFFTileSize(&tif));
+  if (status_or_long.value() != 24576) {
+    return absl::InternalError(
+        absl::StrCat("tiles are ", status_or_long.value(), " bytes\n"));
+  }
+  tsize_t sz = status_or_long.value();
+
+  sapi::v::Array<uint8_t> buffer_(sz);
+  SAPI_ASSIGN_OR_RETURN(
+      status_or_long, api.TIFFReadEncodedTile(&tif, 9, buffer_.PtrBoth(), sz));
+  if (status_or_long.value() != sz) {
+    return absl::InternalError(absl::StrCat(
+        "Did not get expected result code from TIFFReadEncodedTile(): ",
+        status_or_long.value(), " instead of ", sz));
+  }
+
+  unsigned pixel_status = 1;
+  if (status = CheckCluster(0, buffer_, kCluster0); !status.ok()) {
+    LOG(ERROR) << "CheckCluster failed:\n" << status.ToString();
+  }
+  pixel_status &= status.ok();
+
+  if (status = CheckCluster(64, buffer_, kCluster64); !status.ok()) {
+    LOG(ERROR) << "CheckCluster failed:\n" << status.ToString();
+  }
+  pixel_status &= status.ok();
+
+  if (status = CheckCluster(128, buffer_, kCluster128); !status.ok()) {
+    LOG(ERROR) << "CheckCluster failed:\n" << status.ToString();
+  }
+  pixel_status &= status.ok();
+
+  if (!pixel_status) {
+    return absl::InternalError("unexpected pixel_status value");
+  }
+
+  SAPI_ASSIGN_OR_RETURN(
+      status_or_int,
+      api.TIFFSetFieldU1(&tif, TIFFTAG_JPEGCOLORMODE, JPEGCOLORMODE_RGB));
+  if (!status_or_int.value()) {
+    return absl::InternalError("TIFFSetFieldU1 not available");
```

Comment:
When does this happen? Is this code from the time where we had some linking issues? I don't think you need to run under the assumption that this function does not exist as we control how it is built.

Reply:
i think it would return from the previous line if it was a linking issue. It is there because if the TIFFSetField failed to execute successfully. There is no Exception but the return value is unexpected. As an example, if it's not ok to change tag. as an example: "Cannot modify tag while writing"

Reply:
Okay but then the message "TIFFSetFieldU1 **not available**" is incorrect, please rework this :)